### ISSUE TITLE
Decide on format for depth-N pushes

### ISSUE THREAD

**Issue by evert** [evert/push-please#4]
Probably depends on #2.

Currently we have

* A custom format
* A css-inspired format
* A SparQL-inspired format
* A S-Expression format

If we don't want depth-N and only care about one level, we need to just use a structured-header list as this is the way to go for new HTTP headers.

**Comment by gabesullice** [comment]
I would very much like the format to be as simple as possible. I think that means we need to leave a rich query language off the table.

This is why I proposed and fragment-based approach in https://github.com/inadarei/draft-prefer-transclude/issues/5#issuecomment-439972424

A fragment-based approach has the advantage of being completely detached from the media type, being very natural to the language of HTTP, URIs and Web Links, and can support depth-n gracefully.

I'll reproduce my suggestion from that issue here:

```
Prefer: transclude="#/data/0/relationships/writer self"
                    \___________________________/ \__/
                                   |                |
                                fragment      link-relation
```

The fragment identifies the _context_ of the link, not the link itself. The document that this would apply to might look like this:

```
{
  "data": [{
    "type": "article",
    "id": 1,
    "relationships": {
      "writer": {
        "links": {
          "self": "/users/1"
        }
      }
    }
  }, {
    ...
  }]
}
```

And if `users` have profile photos, depth-n could be supported like so:

`Prefer: transclude="(#/data/0/relationships/writer self (#/data/links/profile-photo enclosure))"`

@evert has brought up that `HAL` only has top-level links, so fragments may not be necessary. `_embedded` was the counterpoint, but @evert believes that these would disappear if this spec were available (I think?). I don't think we should ignore nested links though.

The examples I give above are from the a JSON:API formatted response, which is the use-case I'm most interested in. JSON:API _does_ have nested links, so if this spec doesn't support them, it would be much less valuable to me.

Long story short, even if we think this spec will make nested links an anti-pattern, I don't think we should assume that they don't exist. We'll need this spec to see wide-adoption _before_ APIs start to be designed around it, which means supporting APIs as they are, not as they should be.

----

Pros:

- It's easy to disambiguate the selector (i.e. "fragment") because it must be preceded by a hash symbol (`#`). This is always followed by a single space and a link relation.
- Depth-1 can simply elide the parenthesis.
- We can require surrounding parenthesis for depth > 1 which should make it easier to parse.
- Using a fragment means this spec is not implicitly better for any media-type (IOW, we avoid the risk being JSON/HTML-centric)
- Using a fragment leaves room for experimentation. IOW, we punt on the decision to use something like JSON Pointer vs XPath vs CSS selectors vs SPARQL vs ... (my example uses JSON Pointer after the `#` mark, but it could use something else too)

What are the cons people can see?

**Comment by evert** [comment]
I can see why some people would want this, the `Link:` header does have an `anchor` parameter after all.

However, fragments don't replace the need for arbitrary-depth pushes. I think they're a different feature altogether. Maybe it's best to open a new issue for this.

It would also be helpful to get a gauge to figure out who else wants this or would oppose it.

**Comment by gabesullice** [comment]
> However, fragments don't replace the need for arbitrary-depth pushes. I think they're a different feature altogether. Maybe it's best to open a new issue for this.

I'm having a hard time understanding how one can be decided upon without the other.

I see the two requirements as: (1) selecting links _in_ a response (2) selecting links _across_ responses.

S-expressions can work without addressing the first requirement, (i.e. using link relations only): `(author (enclosure))`. The CSS-based approach seems to primarily address only the first requirement. I think SPARQL addresses both at the same time (I haven't dug into it deeply)?

Am I looking at this incorrectly? What am I missing?

**Comment by evert** [comment]
Hi @gabesullice , the syntax proposals for both SparQL and CSS and S-expressions all did more or less the same thing, just expressed with different syntax.

I think my point was that it would be helpful to have a separate discussion on if people wants the ability to pushing by anchor/fragment. If the answer is "yes", then I agree: the syntax we choose will have to support all these use-cases.

So I'm really just suggesting, lets separate the anchor feature discussion from the syntax discussion. It's all too easy to focus and argue on 'how', when the discussion we should be having is 'what'.

**Comment by gabesullice** [comment]
@evert, created that issue: https://github.com/evert/push-please/issues/6

Feel free to edit/add to/simplify as you see fit!

---

If it wasn't already clear, my vote is for a nested parenthesis format: `(foo (bar))`

I think it's easy to parse and human friendly. I do have a question about it though, what's the proposed order of evaluation? Does the outermost group apply to the n=1 response or is it the innermost group? IOW, does `(foo (bar))` mean "please push the `foo` link, then push the `bar` link in that response" or does it mean "please push the `bar` link, then push the `foo` link in that response".

**Comment by asbjornu** [comment]
As this discussion now [surfaced on Slack](https://httpapis.slack.com/archives/C1JQJUF2T/p1559740598002000), I'd like to echo my thoughts there in a less ephemeral place.

Perhaps we could define a subset (or profile) of CSS just including [Type selectors](https://www.w3.org/TR/selectors-3/#type-selectors), [Descendant combinators](https://www.w3.org/TR/selectors-3/#descendant-combinators) and [Grouping](https://www.w3.org/TR/selectors-3/#grouping)? That might get us within the 80% sweet spot @dret is asking for.

Why I'm touting the CSS Selector horn is because the simplest thinkable syntax for transclusion is valid CSS Selector syntax out of the box. Given the following example, I'll show what I mean:

```json
{
    "name": "Han Solo",
    "height": 1.85,
    "friends": [
        { "id": "/characters/r2-d2" },
        { "id": "/characters/luke-skywalker" },
        { "id": "/characters/leia-organa" }
    ],
    "enemies": [
        { "id": "/characters/darth-vader" },
        { "id": "/characters/boba-fett" },
        { "id": "/characters/jabba-the-hutt" }
    ]
}
```

If we want to transclude Han Solo's `friends`, we'll simply use the [type selector](https://www.w3.org/TR/selectors-3/#type-selectors) and write:

```http
Prefer-Push: friends
```

If we want to transclude both of Han Solo's `friends` and `enemies`, we'll use the [grouping selector](https://www.w3.org/TR/selectors-3/#grouping) (`,`):

```http
Prefer-Push: friends, enemies
```

If we want to transclude the `enemies` of Han Solo's `friends`, we can use the [descendant combinator](https://www.w3.org/TR/selectors-3/#descendant-combinators) (` `):

```http
Prefer-Push: friends enemies
```

It's easy to envision expanding this profile of CSS with [Child combinators](https://www.w3.org/TR/selectors-3/#child-combinators), [Universal selectors](https://www.w3.org/TR/selectors-3/#universal-selector), etc., but I think we can get a long way by just using the three components demonstrated above.
